File: 2019/05/rs/src/main.rs
use std::cell::RefCell;

pub struct IntCodeComputer {
    memory: RefCell<Vec<i32>>,
    input: i32,
    output: i32,
    pointer: usize,
    running: bool,
}

impl IntCodeComputer {
    pub fn new(memory: &Vec<i32>, input: i32) -> IntCodeComputer {
        IntCodeComputer {
            memory: RefCell::new(memory.clone()),
            input,
            output: 0,
            pointer: 0,
            running: true
        }
    }

    pub fn run(&mut self) -> i32 {
        while self.running {
            self.tick();
        }
        self.output
    }

    fn get_parameter(&self, offset: i32, mode: i32) -> i32 {
        let value = self.memory.borrow()[self.pointer + offset as usize];
        match mode {
            0 => self.memory.borrow()[value as usize], // POSITION
            1 => value,                                // IMMEDIATE
            _ => {
                panic!("Unrecognized parameter mode '{}'", mode);
            }
        }
    }

    fn get_address(&self, offset: i32) -> usize {
        self.memory.borrow()[self.pointer + offset as usize] as usize
    }

    fn tick(&mut self) {
        if !self.running {
            return
        }
        let instruction = self.memory.borrow()[self.pointer];
        let (opcode, p1_mode, p2_mode) =  (instruction % 100, (instruction / 100) % 10, (instruction / 1000) % 10);
        match opcode {
            1 => { // ADD
                let address = self.get_address(3);
                self.memory.borrow_mut()[address] = self.get_parameter(1, p1_mode) + self.get_parameter(2, p2_mode);
                self.pointer += 4;
            },
            2 => { // MUL
                let address = self.get_address(3);
                self.memory.borrow_mut()[address] = self.get_parameter(1, p1_mode) * self.get_parameter(2, p2_mode);
                self.pointer += 4;
            },
            3 => { // INPUT
                let address = self.get_address(3);
                self.memory.borrow_mut()[address] = self.input;
                self.pointer += 2;
            },
            4 => { // OUTPUT
                self.output = self.get_parameter(1, p1_mode);
                self.pointer += 2;
            },
            5 => { // JMP_TRUE
                if self.get_parameter(1, p1_mode) != 0 {
                    self.pointer = self.get_parameter(2, p2_mode) as usize;
                } else {
                    self.pointer += 3;
                }
            },
            6 => { // JMP_FALSE
                if self.get_parameter(1, p1_mode) == 0 {
                    self.pointer = self.get_parameter(2, p2_mode) as usize;
                } else {
                    self.pointer += 3;
                }
            },
            7 => { // LESS_THAN
                let address = self.get_address(3);
                self.memory.borrow_mut()[address] = if self.get_parameter(1, p1_mode) < self.get_parameter(2, p2_mode) { 1 } else { 0 };
                self.pointer += 4;
            },
            8 => { // EQUALS
                let address = self.get_address(3);
                self.memory.borrow_mut()[address] = if self.get_parameter(1, p1_mode) == self.get_parameter(2, p2_mode) { 1 } else { 0 };
                self.pointer += 4;
            },
            99 => { // HALT
                self.running = false;  
            },
            _ => panic!("Unknown instruction {} {}", self.pointer, opcode)
        }
    }
}

fn solve(memory: &Vec<i32>) -> (i32, i32) {
    (
        IntCodeComputer::new(memory, 1).run(),
        IntCodeComputer::new(memory, 5).run()
    )
}

fn get_input(file_path: &String) -> Vec<i32> {
    std::fs::read_to_string(file_path).expect("Error reading input file!")
        .split(',')
        .map(|i| i.parse().unwrap())
        .collect()
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() != 2 {
        panic!("Please, add input file path as parameter");
    }
    let now = std::time::Instant::now();
    let (part1_result, part2_result) = solve(&get_input(&args[1]));
    println!("P1: {}", part1_result);
    println!("P1: {}", part2_result);
    println!();
    println!("Time: {:7}", now.elapsed().as_secs_f32());
}

File: 2019/05/c/main.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/time.h>
#include <string.h>

typedef struct
{
    int *memory, size, capacity;
} Input;
typedef struct
{
    int part1;
    int part2;
} Results;

typedef struct
{
    int *memory;
    int pointer;
    int running;
    int input;
    int output;
} IntCodeComputer;

char errorMessage[64];

int getParameter(IntCodeComputer *computer, int offset, int mode)
{
    int value = computer->memory[computer->pointer + offset];
    switch (mode)
    {
    case 0: // POSITION
        return computer->memory[value];
    case 1: // IMMEDIATE
        return value;
    }
    sprintf(errorMessage, "Unrecognized parameter mode '%d'\n", mode);
    perror(errorMessage);
    exit(1);
}

int getAddress(IntCodeComputer *computer, int offset)
{
    return computer->memory[computer->pointer + offset];
}

void tick(IntCodeComputer *computer)
{
    if (!computer->running)
        return;
    int instruction = computer->memory[computer->pointer];
    int opcode = instruction % 100, p1Mode = (instruction / 100) % 10, p2Mode = (instruction / 1000) % 10;
    switch (opcode)
    {
    case 1: // ADD
        computer->memory[getAddress(computer, 3)] = getParameter(computer, 1, p1Mode) + getParameter(computer, 2, p2Mode);
        computer->pointer += 4;
        break;
    case 2: // MUL
        computer->memory[getAddress(computer, 3)] = getParameter(computer, 1, p1Mode) * getParameter(computer, 2, p2Mode);
        computer->pointer += 4;
        break;
    case 3: // INPUT
        computer->memory[getAddress(computer, 1)] = computer->input;
        computer->pointer += 2;
        break;
    case 4: // OUTPUT
        computer->output = getParameter(computer, 1, p1Mode);
        computer->pointer += 2;
        break;
    case 5: // JMP_TRUE
        if (getParameter(computer, 1, p1Mode))
            computer->pointer = getParameter(computer, 2, p2Mode);
        else
            computer->pointer += 3;
        break;
    case 6: // JMP_FALSE
        if (!getParameter(computer, 1, p1Mode))
            computer->pointer = getParameter(computer, 2, p2Mode);
        else
            computer->pointer += 3;
        break;
    case 7: // LESS_THAN
        computer->memory[getAddress(computer, 3)] = getParameter(computer, 1, p1Mode) < getParameter(computer, 2, p2Mode)
                                                        ? 1
                                                        : 0;
        computer->pointer += 4;
        break;
    case 8: // LESS_THAN
        computer->memory[getAddress(computer, 3)] = getParameter(computer, 1, p1Mode) == getParameter(computer, 2, p2Mode)
                                                        ? 1
                                                        : 0;
        computer->pointer += 4;
        break;
    case 99: // HALT
        computer->running = 0;
        break;
    default:
        sprintf(errorMessage, "Unknow instruction '%d' at '%d'\n", opcode, computer->pointer);
        perror(errorMessage);
        exit(1);
        break;
    }
}

IntCodeComputer newComputer(Input memory, int input)
{
    int *newMemory = malloc(memory.size * sizeof(int));
    memcpy(newMemory, memory.memory, memory.size * sizeof(int));
    return (IntCodeComputer){
        newMemory,
        0,
        1,
        input,
        0};
}

int run(IntCodeComputer *computer)
{
    while (computer->running)
        tick(computer);
    free(computer->memory);
    return computer->output;
}

int runProgram(Input memory, int input)
{
    IntCodeComputer computer = newComputer(memory, input);
    return run(&computer);
}

Results solve(Input input)
{
    return (Results){runProgram(input, 1), runProgram(input, 5)};
}

#define INPUT_INCREMENT 10
void addToInput(Input *input, int value)
{
    if (input->size == input->capacity)
    {
        input->capacity += INPUT_INCREMENT;
        input->memory = realloc(input->memory, input->capacity * sizeof(int));
    }
    input->memory[input->size++] = value;
}

Input getInput(char *filePath)
{
    FILE *file;
    if ((file = fopen(filePath, "r")) == NULL)
    {
        perror("Error reading input file!\n");
        exit(1);
    }
    fseek(file, 0, SEEK_END);
    long length = ftell(file);
    rewind(file);
    char *content = malloc(length);
    fread(content, 1, length, file);
    Input input = {
        calloc(INPUT_INCREMENT, sizeof(int)),
        0,
        INPUT_INCREMENT};
    char *value = strtok(content, ",");
    while (value != NULL)
    {
        addToInput(&input, atoi(value));
        value = strtok(NULL, ",");
    }
    fclose(file);
    return input;
}

void freeInput(Input input)
{
    free(input.memory);
}

int main(int argc, char **argv)
{
    if (argc != 2)
    {
        perror("Please, add input file path as parameter");
        exit(1);
    }
    struct timeval starts, ends;
    gettimeofday(&starts, NULL);
    Input input = getInput(argv[1]);
    Results results = solve(input);
    gettimeofday(&ends, NULL);
    freeInput(input);
    printf("P1: %d\n", results.part1);
    printf("P2: %d\n\n", results.part2);
    printf("Time: %.7f\n", (double)((ends.tv_sec - starts.tv_sec) * 1000000 + ends.tv_usec - starts.tv_usec) / 1000000);
    return 0;
}
